camel/societies/role_playing.py

```diff
@@ -514,8 +532,158 @@ async def ainit_chat(
             content=init_msg_content,
         )
 
+        # Reset conversation turn counter
+        self._conversation_turn = 0
+
         return init_msg
 
+    def _create_checkpoint(self, checkpoint_name: str) -> None:
+        r"""Create a memory checkpoint by saving the current state of both
+            agents' memories.
+
+        Args:
+            checkpoint_name (str): Name to identify this checkpoint
+        """
+        # Create deep copies of agent memories to preserve their state
+        assistant_memory_records = [
+            record.memory_record
+            for record in self.assistant_agent.memory.retrieve()
+        ]
+        user_memory_records = [
+            record.memory_record
+            for record in self.user_agent.memory.retrieve()
+        ]
+
+        # Ensure there are user messages before continuing
+        if not user_memory_records:
+            logger.warning("Cannot create checkpoint : No user messages")
+            return
+
+        last_user_message = user_memory_records[-1].message.content
+
+        # Check if a checkpoint with this name already exists
+        if checkpoint_name in self._memory_checkpoints:
+            logger.info(f"Overwriting existing checkpoint '{checkpoint_name}'")
+            # Preserve the original checkpoint's turn number
+            turn = self._memory_checkpoints[checkpoint_name].get(
+                "turn", self._conversation_turn
+            )
+        else:
+            turn = self._conversation_turn
+
+        self._memory_checkpoints[checkpoint_name] = {
+            "assistant": copy.deepcopy(assistant_memory_records),
+            "user": copy.deepcopy(user_memory_records),
+            "turn": turn,
+            "timestamp": datetime.now().timestamp(),
+            "user_message": last_user_message,
+        }
+
+        logger.info(f"Created checkpoint '{checkpoint_name}' at turn {turn}")
+
+    def list_checkpoints(self) -> List[Dict]:
```

Comment:
more specific type hint

```suggestion
    def list_checkpoints(self) -> List[Dict[str, Union[str, int, float]]]:

```